common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -513,12 +511,17 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
       msg.shuffleId, msg.shuffleMergeId, msg.appId, msg.appAttemptId);
     AppShuffleInfo appShuffleInfo = validateAndGetAppShuffleInfo(msg.appId);
     if (appShuffleInfo.attemptId != msg.appAttemptId) {
-      // If this Block belongs to a former application attempt, it is considered late,
-      // as only the blocks from the current application attempt will be merged
-      // TODO: [SPARK-35548] Client should be updated to handle this error.
+      // If finalizeShuffleMerge from a former application attempt, it is considered late,
+      // as only the finalizeShuffleMerge request from the current application attempt will be merged.
+      // Too old app attempt only being seen by an already failed app attempt, and no need use callback to
+      // return to client now, because the finalizeShuffleMerge in DAGScheduler has no retry policy,
+      // and don't use the BlockPushNonFatalFailure because it's the finalizeShuffleMerge related case,
+      // not the block push case, just throw it in server side now.
+      // TODO we may use a new exception class to include the finalizeShuffleMerge related case
+      // just as the BlockPushNonFatalFailure contains the block push cases.
```

Comment:
@Victsm @Ngone51 @mridulm 
About the finalizeShuffleMerge case, i think it's fine to keep it now just throw in server side, in future if we have retry policy or other cases, we can add a class such as BlockPushNonFatalFailure, and now it's only about the old attempt case without retry.
What's your opinion?
Thanks.

Reply:
I think for now keeping the current behavior is good since it simplifies things.
We can see how to better handle it if we have the need of retry in the future.

Reply:
Thanks @Victsm for your reply and confirm!

Reply:
@zhuqi-lucas  Oh, by "revert", I meant that we don't need to handle the `BlockPushNonFatalFailure` specifically but throwing the `BlockPushNonFatalFailure` here is fine.

Reply:
I'm OK with using `IllegalArgumentException` since we are not doing any special handling for when this exception gets thrown during shuffle merge finalization either on the server side or client side.